src/bridge/test-router.c

```diff
@@ -264,6 +294,123 @@ test_external_bridge (TestCase *tc,
   g_object_unref (shim_transport);
 }
 
+static void
+test_external_channel_redirect (TestCase *tc,
+                                gconstpointer unused)
+{
+  CockpitRouter *router;
+  GBytes *sent;
+  CockpitPeer *peer, *other_peer;
+  JsonObject *other_peer_config, *other_peer_match, *control;
+
+  static CockpitPayloadType payload_types[] = {
+    { "echo", mock_echo_channel_get_type },
+    { NULL },
+  };
+
+  /* Create configuration for second peer based on the mock config with --upper replaced with --lower */
+  other_peer_match = json_object_new ();
+  json_object_set_string_member (other_peer_match, "payload", "lower");
+  other_peer_config = json_object_new ();
+  for (GList *elem = json_object_get_members (tc->mock_config); elem; elem = elem->next)
+    {
+      const gchar *key = elem->data;
+      if (g_strcmp0 (key, "spawn") == 0)
+        {
+          JsonArray *old_argv = json_node_get_array (json_object_get_member (tc->mock_config, key));
+          JsonArray *new_argv = json_array_new ();
+          for (GList *arr_elem = json_array_get_elements (old_argv); arr_elem; arr_elem = arr_elem->next)
+          {
+            const gchar *arr_data = json_node_get_string (arr_elem->data);
+            if (g_strcmp0 (arr_data, "--upper") == 0)
+              json_array_add_string_element (new_argv, "--lower");
+            else
+              json_array_add_string_element (new_argv, g_strdup (arr_data));
+          }
+          json_object_set_array_member (other_peer_config, g_strdup (key), new_argv);
```

Comment:
That's one of the leaks -- [json_object_set_array_member](https://developer.gnome.org/json-glib/stable/json-glib-JSON-Object.html#json-object-set-array-member) does not take ownership of `key`, it dups it internally. So just drop the g_strdup().